src/fetch.rs

```diff
@@ -0,0 +1,165 @@
+// use error_chain::error_chain;
+use color_eyre::{
+    eyre::{eyre, WrapErr},
+    Result,
+};
+use log::{info, trace};
+use regex::Regex;
+use reqwest::header::{HeaderValue, CONTENT_LENGTH, RANGE};
+use reqwest::StatusCode;
+use std::str::FromStr;
+
+// error_chain! {
+//     foreign_links {
+//         Io(std::io::Error);
+//         Reqwest(reqwest::Error);
+//         Header(reqwest::header::ToStrError);
+//     }
+// }
+
+struct PartialRangeIter {
+    start: usize,
+    end: usize,
+    buffer_size: usize,
+}
+
+impl PartialRangeIter {
+    pub fn new(start: usize, end: usize, buffer_size: usize) -> Result<Self> {
+        if buffer_size == 0 {
+            Err(eyre!("Invalid buffer size, give a value greater than 0"))?;
+        }
+        Ok(PartialRangeIter {
+            start,
+            end,
+            buffer_size,
+        })
+    }
+}
+
+impl Iterator for PartialRangeIter {
+    type Item = HeaderValue;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.start >= self.end {
+            None
+        } else {
+            let prev_start = self.start;
+            self.start += std::cmp::min(self.buffer_size as usize, self.end - self.start + 1);
+            Some(
+                HeaderValue::from_str(&format!("bytes={}-{}", prev_start, self.start - 1))
+                    .wrap_err_with(|| eyre!("Failed to create header value"))
+                    .unwrap(),
+            )
+        }
+    }
+}
+
+pub async fn retrieve_file(
+    list: &crate::repo::Wordlist,
+    decompress: bool,
+    base_dir: &str,
+    user_agent: &str,
+    worker_count: usize,
+) -> Result<()> {
+    assert!(
+        !std::path::Path::new(&format!("/{}/{}", base_dir, list.get_name()))
+            .try_exists()
+            .expect("File already exists")
+    );
```

Comment:
This makes it necessary to specify base-dir as an absolute path. Furthermore, the entire download aborts because the error is not handled in main.
```suggestion
    let file = Path::new(base_dir).join(list.get_name());
    if file.exists() {
        info!("{} already exists -- skipping", list.get_name());
        return Ok(());
    }
```

Reply:
And the python version created subdirectories for the groups. We should keep the file structure as it was.